README.md
<!-- @import "[TOC]" {cmd="toc" depthFrom=3 depthTo=4 orderedList=false} -->

<!-- code_chunk_output -->

- [零、下载说明](#-零-下载说明-)
- [一、框架简介](#-一-框架简介-)
- [二、目录结构](#-二-目录结构-)
- [三、基本概念](#-三-基本概念-)
  - [(1) view](#-1-view-)
  - [(2) ctrl](#-2-ctrl-)
  - [(3) controller](#-3-controller-)
  - [(4) service](#-4-service-)
  - [(5) page](#-5-page-)
  - [(6) router](#-6-router-)
  - [(7) app](#-7-app-)
  - [(8) lgui entry](#-8-lgui-entry-)
  - [(9) lgui c library](#-9-lgui-c-library-)
  - [(10) lua table](#-10-lua-table-)
- [四、lgui framework文件说明](#-四-lgui-framework文件说明-)
- [五、时序图](#-五-时序图-)
  - [(1) lgui framework starting](#-1-lgui-framework-starting-)
  - [(2) page loading](#-2-page-loading-)
  - [(3) service callback](#-3-service-callback-)
- [六、lgui framework特性描述](#-六-lgui-framework特性描述-)
  - [1、ctrl加载顺序](#-1-ctrl加载顺序-)
  - [2、setdata特性](#-2-setdata特性-)
  - [3、view中ctrl事件关联controller](#-3-view中ctrl事件关联controller-)
  - [4、this特性](#-4-this特性-)
  - [5、热更新特性](#-5-热更新特性-)
  - [6、lgui中的背景特性](#-6-lgui中的背景特性-)
  - [7、ctrl position特性](#-7-ctrl-position特性-)
  - [8、图片资源自动解析特性](#-8-图片资源自动解析特性-)
  - [9、ctrl的父子特性](#-9-ctrl的父子特性-)
  - [10、service抢占默认事件特性](#-10-service抢占默认事件特性-)
  - [11、文本重新着色特性](#-11-文本重新着色特性-)
  - [12、common页面特性](#-12-common页面特性-)
  - [13、手势导航特性](#-13-手势导航特性-)
- [七、函数说明](#-七-函数说明-)
  - [1、lua framework API](#-1-lua-framework-api-)
  - [2、c library API <lgui3.h>](#-2-c-library-api-lgui3h-)
- [八、控件说明](#-八-控件说明-)
  - [1、bar](#-1-bar-)
  - [2、btn](#-2-btn-)
  - [3、btnm](#-3-btnm-)
  - [4、checkbox](#-4-checkbox-)
  - [5、clock](#-5-clock-)
  - [6、img](#-6-img-)
  - [7、list](#-7-list-)
  - [8、loading](#-8-loading-)
  - [9、mark](#-9-mark-)
  - [10、mbox](#-10-mbox-)
  - [11、text](#-11-text-)
  - [12、roller](#-12-roller-)
  - [13、slider](#-13-slider-)
  - [14、spinner](#-14-spinner-)
  - [15、page](#-15-page-)
  - [16、textarea](#-16-textarea-)
  - [17、qrcode](#-17-qrcode-)
  - [18、tab](#-18-tab-)
  - [19、vdec](#-19-vdec-)
  - [20、blank](#-20-blank-)
  - [21、swiper](#-21-swiper-)
  - [22、shared](#-22-shared-)
  - [23、chart](#-23-chart-)
  - [24、combobox](#-24-combobox-)
  - [25、drawer](#-25-drawer-)
  - [26、aimg](#-26-aimg-)
  
<!-- /code_chunk_output -->

<script>

ChangeLog
2023-02-27
1. 处理只缓存单张图片时,出现的异常问题.
2. 修复缩放图片多次被监听的问题.
3. 图片文件后缀不区分大小写。
4. 添加设置 cache_size_max 的接口。
5. 修复list初始select参数未设置时,点击会界面卡死的问题。
6. 修复屏保时间从0设置为其他时,未生效的问题.
7. 修复btnm控件set_data时死机的BUG
8. btnm 控件添加 pad_x, pad_y 用于调节按键背景的填充宽度以及高度
9. 处理特殊情况下颜色混合时,坐标偏移的问题
10. 优化存在动画页面, 销毁common页面慢的问题
11. 修复存在动画页面, drawer 控件上按键效果显示异常的问题
12. list 添加 btn_fit 属性, 用于设置选项按键的宽高的fit属性
13. slider 控件增加 res_knob 属性. 可将旋钮替换成图片。
14. list 子控件 TEXT 添加 pad_x, pad_y 用于设置左右,上下的填充长度
15. 生效aimg控件set_data设置times属性时,重置循环次数
16. btnm 添加属性 inner,用于调节按键之间的距离
17. clock 控件添加属性 hour_offset_x,hour_offset_y,min_offset_x,min_offset_y,sec_offset_x,sec_offset_y 用于调整时钟图片的旋转中心
18. 添加公共属性 dis_scroll_propagation, 用于禁用当前控件的滑动传递功能
19. list 子控件添加 pad_x, pad_y 用于调节文本背景的填充宽度以及高度
20. page 控件添加属性 color_edge 用于设置控件到达最高或最低位置时,显示类似圆圈效果的颜色

2022-05-18:
1. 修复在父tab控件上直接创建list和page控件时,销毁页面报错的问题.
2. 增加list控件属性:
   radius_btn: list内部按键的圆角半径。
   select: list单选模式下,当前选中的选项.(取值从1开始)
   single: 是否为单选模式 (true:单选 false:多选)

2022-04-27:
1. 修复缩放jpeg图片右边会有马赛克的问题
2. 修复程序启动时UI有概率不显示的问题
3. 修复list imgbtn与btn同时使用时imgbtn显示错误的问题
4. 添加qrcode 容错属性 eclevel (0-3)
5. 添加list中图片按键以及空白按键的长按事件
6. 修复缩放高度小于2的图片时系统报错的问题
7. 修复list父类是mark时滑动报错的问题
8. 控件list添加属性 radius 背景圆角
9. 添加page控件之间的滑动传递, 只有page的父控件是page、tab 或 drawer 时生效。
10. blank 添加自动匹配宽高的功能,通过w,h的特殊值设置控件的自动匹配模式. 详细数值查看文档。
11. blank 颜色属性以及圆角属性。

2022-03-15:
1. 修复textarea控件在密码模式文本过长时显示异常的问题

2022-03-14:
1. 修复自定义定时器删除之后再创建时，定时器时序不对的问题

2022-03-10:
1. 添加属性 c_bg  text控件背景颜色
2. 添加属性 radius  text控件背景圆角
3. 添加属性 anim_time tab控件切换到下个tab所用的动画时间
4. 添加属性 treshold  tab控件滑动时切换到下个界面的阈值(指最少拖动treshold距离切换到下个tab)
5. 修复text控件mode属性为MODE_EXPAND时,居中异常的问题
6. 增加c层开关显示的接口 lgui_dispaly_deinit() lgui_dispaly_reinit()
7. 修复aimg控件循环播放时最后一张图片时间过长的问题
8. lua 添加 lgui_gesture_set_part(part) 接口，用于设置手势可触发的部分，可取值参考 utils/gesture.lua
9. 修复page设置 slideposy 时, 子控件使用ref且ref==parent时,显示异常的问题
10. 机器上显示时, 背景会根据屏幕大小自动缩放铺满屏幕
11. 添加drawer控件的属性 dir, 用于设置drawer控件的滑动方向, 可取值参考 utils/drawer.lua

2022-03-03:
1. 修复page控件禁用且子类存在btn时,滑动不能传递给父类tab的问题
2. 修复当tab控件有mark子类, 在mark控件的范围内滑动tab时,mark控件的up事件会触发的问题
3. 修复界面存在spnnier控件时onshow以及timer不会回调的问题。
4. 更新pc端lgui3库, 优化pc图片缩放, 可任意比例缩放。

2022-02-24:
1. 支持阿拉伯文显示。
2. 增加 textarea 光标颜色设置。处理set_data光标闪烁周期属性,不生效问题。
3. 增加 common_add 函数能携带到common界面的msg参数
4. 修复 btnm 中map属性为 nil 时, 程序异常
5. 增加 page 的属性 slideposx, slideposy。用于设置page页面的起始位置
6. 增加 list 的long事件
7. 优化 aimg 控件的显示效果
8. 增加手动设置屏保时间的函数
9. 处理使用list TYPE_BTN 类型时退出界面死机的问题。
10. 增加 page 的inner属性, 用于设置子对象之间的内边距,(layout ~= utils_page.LAYOUT_OFF时)。
11. 更新PC模拟器的lgui3动态库。
12. 修复 slider 控件 w > h 且半透明时, 中间镂空, 两边有颜色的问题。
13. 处理delete drawer控件时, 段错误的问题
14. 添加 btn、blank、mark、mbox 控件的 lost 事件
15. 添加 坐标点上报。事件名为 lgui_event_touch。参数为 x:x坐标 y:y坐标 state:触摸状态
16. 修复mark颜色属性 set_data 失败的问题。
17. list 添加imgbtn类型。
18. 可触发控件增加touchable属性,用于禁用对象。
19. mark增加 radius 属性设置圆角半径。
20. 增加自定义定时器的接口。 lgui_timer_create(info)、lgui_timer_delete()
                         lgui_timer_pause()、 lgui_timer_reumse()。 详情请查看说明文档。
21. 添加 list 控件的属性, list控件支持横向滑动
22. 增加btn以及list按下效果的延时处理。
23. 修复一些问题。

2021-05-28:
1. 增加WIN64模拟器模式
2. 运行模拟器需修改config/env.lua的path和cpath，其中cpath应修改为winlib
3. 修改后，用户使用run_simulator.bat即可运行模拟器
4. 鼠标左键点击即模拟触摸屏点击操作
5. LGUI3的page、controller等代码和界面效果，模拟器和嵌入式完全通用，无缝衔接
6. 用户迁移工程至嵌入式平台时，可以删除winlib和lua.exe等模拟器运行需要的组件

2021-05-08:
1. 处理page控件的子控件set_data时, 坐标异常的问题
2. env.lua 增加设置screensaver_time:无触摸进入屏保的超时时间(s)。当触发屏保时，会触发service: lgui_event_screensaver。
   屏保由用户自行实现,框架只产生service。
   触摸或调用lgui_screensaver_reset() 都会重新计算进入屏保的时间。
3. 增加手势导航特性。通过从边缘向内滑动响应手势事件，支持左、右、上滑。默认不开启，透明颜色。
   使用函数 lgui_gesture_open() 开启手势导航。
   使用函数 lgui_gesture_close() 关闭手势导航。
   使用函数 lgui_gesture_set_color(color) 设置颜色。
4. 将NX5的视频库更新到 libVideoDec-1.6.6.so。
5. 修复一些问题。

2021-04-17:
1. 增加 config/env.lua 中 cache_max_size 配置。
2. 增加 img 控件属性，用于只显示图片的部分区域。
3. 增加 text 控件属性，用于设置控件高度。
4. 增加 list 控件属性，用于显示控件到达最高或最低时的动画效果。
5. 增加 aimg 控件，使用多张图片达到显示动图的效果。
6. lgui_push_table 增加将数组压栈。将 key 设置为 0 即可. 具体示例请看README.md.
7. 处理已知BUG

2021-03-27:
1. 增加btn控件的res_long属性, 为长按时显示的图片效果
2. 增加swiper控件的mode属性, 增加循环模式。值参考:utils/swiper.lua.
3. 修改swiper控件触发方式, 只有change回调返回选中成员的编号，其他事件不返回成员编号，只返回事件
4. 修复若干BUG

2021-02-28:
1. 添加控件 combobox。详情请查看文档。
2. 处理lgui_push_table函数中value_type为LUA_TSTRING时，value传数组程序崩溃的问题。
   当value_type为LUA_TSTRING时, value不需要对字符串或数组的地址取指针,直接复制即可。其他类型依然需要取地址
   /* example:
   ** LGUI_LUA_API int test_push(lua_State *L)
   ** {
   **    lgui_table_map_s map[4];
   **    char *value1 = "value1";
   **    char value2[] = "value2";
   **    int  value3 = 3;
   **    map[0].key = (int)"key1";
   **    map[0].value = value1;             /* string 类型不用取址 */
   **    map[0].value_type = LUA_TSTRING;
   **    map[1].key = (int)"key2";
   **    map[1].value = value2;             /* string 类型不用取址 */
   **    map[1].value_type = LUA_TSTRING;
   **    map[2].key = (int)"key3";
   **    map[2].value = (char *)&value3;    /* number 类型需要取址 */
   **    map[2].value_type = LUA_TNUMBER;
   **    map[3].value_type = LUA_TNONE;
   **    lgui_push_table(L, map);
   **    return 1;
   ** }
   */
3. 增加控件 drawer。
4. 修复已知BUG

2021-02-06:
1. 增加文本重新着色功能. 例如: "Write a \x1b0000ff blue\x1b word"中blue文本将为蓝色.
   提供 get_recolor_str 函数，实现文本重新着色的字符拼接。
   注: 文本重新作色只能在单行进行。多行文本只着色开始的一行。支持的控件为 text, list。不支持设置alpha值。重新着色的alpha为view中设置的alpha。
       重新着色的文本内部不能再次重新着色。例如:"\x1bff0000 Write a \x1b0000ff blue\x1b word\x1b",这样的方式是不允许的
2. 增加common界面特性, 详情请查看文档
3. 修复已知BUG

2021-01-31:
1. roller、btnm控件的up事件不再有返回值，只有change事件有返回值
2. list 响应事件更改, up事件返回值为点击按键响应, change 事件响应返回值为当前list的滑动位置
3. list 控件增加 slidepos 属性，设置list初始滑动位置
4. 解决使用相对位置的控件，set_data之后位置更改的问题
5. 修复一些问题
6. 修改控件演示示例

2021-01-24:
1. 处理 blank 与 btn 控件在page滑动不灵敏问题
2. page 增加round属性设置圆角
3. 添加chart控件
4. 解决一些问题

2021-01-16:
1. page、list、tab 增加 c_bar 设置右侧以及底部的滚动条颜色,默认不显示
2. 优化图片缩放方案,将图片缩数据加入cache方案
3. 增加 env.ttf_cache_num 配置,用于设置ttf对象缓存的最大数量
4. 修复已知BUG


pages/home.lua
--[=[
/******************************************************************************
Copyright(c) 2011-2021 Digital Power Inc.
File name: home.lua
Author: liuzhengzhong
Version: 1.0.0
Date: 2020/12/7
Description:
History:
Bug report: liuzhengzhong@d-power.com.cn
******************************************************************************/
]=]

local map = {
    "home/消息.png", "01/01", "物业通知：因水管维修，今晚10点小区要停水，请住户注意蓄水",
    "home/消息.png", "03/12", "物业通知：因水管维修，今晚11点小区要停水，请住户注意蓄水",
    "home/消息.png", "04/18", "物业通知：因水管维修，今晚12点小区要停水，请住户注意蓄水",
    "home/消息.png", "05/15", "物业通知：因水管维修，今晚13点小区要停水，请住户注意蓄水",
    "home/消息.png", "06/12", "物业通知：因水管维修，今晚14点小区要停水，请住户注意蓄水",
    "home/消息.png", "06/18", "物业通知：因水管维修，今晚15点小区要停水，请住户注意蓄水",
    "home/消息.png", "07/20", "物业通知：因水管维修，今晚16点小区要停水，请住户注意蓄水",
    "home/消息.png", "10/15", "请10月11日起新疆喀什地区返珠或与该地区有接触人员有接触自觉填报《重点地区来珠人员情况表》，没有到过新疆或无相关接触人员无需填写。谢谢！",
}

local _set_map = {
    "个性设置", "(桌面背景、电子相框等)", "home/list_line.png",
    "音量设置", "(通话音量、铃声音量等)", "home/list_line.png",
    "智能家居", "(网关功能、搜索网关等)", "home/list_line.png",
    "开门密码", "(住户开门密码)",        "home/list_line.png",
}

local _record_map = {
    "门口机",   "2021-01-01 12:15:00", "home/list_line2.png",
    "室内机",   "2021-01-06 08:18:00", "home/list_line2.png",
    "管理机",   "2021-01-16 15:16:00", "home/list_line2.png",
    "管理机",   "2021-01-25 17:16:00", "home/list_line2.png",
    "室内机",   "2021-02-12 15:28:00", "home/list_line2.png",
    "门口机",   "2021-02-18 11:48:00", "home/list_line2.png",
    "门口机",   "2021-02-28 16:13:00", "home/list_line2.png",
    "管理机",   "2021-03-12 20:15:00", "home/list_line2.png",
    "室内机",   "2021-03-19 12:16:00", "home/list_line2.png",
}

local controller =
{
    data =
    {
        list_map = map,
        mbox_title = "",
        mbox_body = "",
        mbox_hidden = true,
        page0_hidden = false,
        page1_hidden = true,
        page2_hidden = true,

        mark0_hidden = true,
        mark1_hidden = true,

        home_set_hidden = false,
        home_text = "",
        record_hidden = true,

        top_tab0_txt_h = 25,
        top_tab1_txt_h = 15,
        top_tab2_txt_h = 15,

        pwd_text = "",

        set_map = _set_map;
        record_map = _record_map;

        page0_tabbottom_left_img0 = "home/轮播1.png",
        page0_tabbottom_left_img1 = "home/轮播2.png",
        page0_tabbottom_left_img2 = "home/轮播2.png",

        page0_tabbottom_right_img0 = "home/轮播1.png",
        page0_tabbottom_right_img1 = "home/轮播2.png",
        page0_tabbottom_right_img2 = "home/轮播2.png",

        chart_map   = {{0xFF385789, 100, 132, 78, 68, 150, 180, 230, 200, 240, 140, 180, 300}},
        chart_map2  = {{0xFF385789, 300, 180, 140, 240, 200, 78, 68, 230, 100, 380, 180, 240}}

    },

    onload = function()
        common_hidden("common")
    end,

    onshow = function()
        log(3, "home onshow")
    end,

    ondestroy = function()
        log(3, "home ondestroy")
    end,

    mbox_action = function(ctrl, value)
        set_data({mbox_hidden = true, mark0_hidden = true})
        if this.data.page0_hidden then
            set_data({page0_hidden = false,  page1_hidden = true,  page2_hidden = true})
        end
    end,

    set_mode = function(v)
        set_data({mark0_hidden = false, mbox_title = "设置模式", mbox_body = "设置" .. v.name .. "成功", mbox_hidden = false})
    end,

    mark_action = function(ctrl)
        set_data({mark0_hidden = true})
    end,

    mark1_action = function(ctrl)
        set_data({mark1_hidden = true})
    end,

    input_set = function(ctrl)
        set_data({mark1_hidden = false, pwd_text = ""})
    end,

    btnm_action = function(ctrl, value)
        if value == "删除" then
            set_data({pwd_text = this.data.pwd_text:sub(1, -2)})
        elseif value == "确定" then
            set_data({mark1_hidden = true, page0_hidden = true, page1_hidden = false,  page2_hidden = true})
        else
            set_data({pwd_text = this.data.pwd_text .. value})
        end
    end,

    ta_change = function(ctrl, value)
        this.data.pwd_text = value:sub(1, 6)
    end,

    call_out = function(ctrl)
        set_data({page0_hidden = true, page1_hidden = true,  page2_hidden = false, home_set_hidden = false, record_hidden = true,home_text = ""})
    end,

    tab_act = function(v, value)
        if v.name == "page0_tabtop" then
            set_data({
                top_tab0_txt_h = 15,
                top_tab1_txt_h = 15,
                top_tab2_txt_h = 15,
                ["top_tab"..value.."_txt_h"] = 25,
            })
        elseif v.name == "page0_tabbottom_left" then
            set_data({
                page0_tabbottom_left_img0 = "home/轮播2.png",
                page0_tabbottom_left_img1 = "home/轮播2.png",
                page0_tabbottom_left_img2 = "home/轮播2.png",
                ["page0_tabbottom_left_img" .. value] = "home/轮播1.png",
            })
        elseif v.name == "page0_tabbottom_right" then
            set_data({
                page0_tabbottom_right_img0 = "home/轮播2.png",
                page0_tabbottom_right_img1 = "home/轮播2.png",
                page0_tabbottom_right_img2 = "home/轮播2.png",
                ["page0_tabbottom_right_img" .. value] = "home/轮播1.png",
            })
        end
    end,

    home_input = function(ctrl, value)
        if value == "删除" then
            set_data({home_text = this.data.home_text:sub(1, -2)})
        elseif value == "呼叫" then
            print("呼叫".. this.data.home_text)
        else
            set_data({home_text = this.data.home_text .. value})
        end
    end,

    home_change = function(ctrl, value)
        this.data.home_text = value
    end,

    show_record = function(ctrl, value)
        set_data({home_set_hidden = true, record_hidden = false,})
    end,

    gesture = function(dir)
        print(dir)
    end
}

return controller
